oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java

```diff
@@ -163,26 +204,37 @@ public String retrieveSubjectToken() throws IOException {
   }
 
   private String retrieveSubjectTokenFromCredentialFile() throws IOException {
-    IdentityPoolCredentialSource identityPoolCredentialSource =
-        (IdentityPoolCredentialSource) credentialSource;
     String credentialFilePath = identityPoolCredentialSource.credentialLocation;
     if (!Files.exists(Paths.get(credentialFilePath), LinkOption.NOFOLLOW_LINKS)) {
       throw new IOException(
           String.format(
               "Invalid credential location. The file at %s does not exist.", credentialFilePath));
     }
     try {
-      return new String(Files.readAllBytes(Paths.get(credentialFilePath)));
+      return parseToken(Files.readAllBytes(Paths.get(credentialFilePath)));
     } catch (IOException e) {
       throw new IOException(
           "Error when attempting to read the subject token from the credential file.", e);
     }
   }
 
-  private String getSubjectTokenFromMetadataServer() throws IOException {
-    IdentityPoolCredentialSource identityPoolCredentialSource =
-        (IdentityPoolCredentialSource) credentialSource;
+  private String parseToken(byte[] token) throws IOException {
+    if (identityPoolCredentialSource.credentialFormatType == CredentialFormatType.TEXT) {
+      return new String(token);
+    }
+
+    JsonFactory jsonFactory = OAuth2Utils.JSON_FACTORY;
+    JsonObjectParser parser = new JsonObjectParser(jsonFactory);
+    GenericJson fileContents =
+        parser.parseAndClose(new ByteArrayInputStream(token), OAuth2Utils.UTF_8, GenericJson.class);
```

Comment:
You can set the object parser on the request and call `response.parseAs(GenericJson.class)` rather than taking the response, pulling into a string, converting to bytes, wrapping with a InputStream to be parsed.

Reply:
I'd still need this for url-sourced credentials though, no?

Reply:
`response.getContent()` will return you an `InputStream` in that case, so the `parseToken` could take an `InputStream`. For the file based json, you could get a `FileInputStream`.